### PR TITLE
Artifact Fetching Module

### PR DESCRIPTION
**Pull request by schuylermartin45**
Introduces the `fetcher/` module to CRM. This module is capable of downloading and extracting build artifacts to a secure temporary directory. This will be coupled with a new `scanner` module for future automation efforts.

Other changes:
- Adds `pyfakefs` as a test dependency (used for mocking file systems)
- Adds a series of `MockHttp*` classes to mock-out responses from `requests`
- Starts work on a `git`-based fetcher
- Starts work on a `from_recipe` factory method that will construct fetcher instances from a recipe's `/source` section.

### CODE REVIEW

**Comment by schuylermartin45** [comment]
@ForgottenProgramme Tagging you for visibility. Let's work together on deprecating the `MockWriter` class in favor of using `pyfakefs`

**Comment by schuylermartin45** [comment]
This will remain in draft-mode until I finish adding unit test coverage for `http_artifact_fetcher.py`

File: conda_recipe_manager/fetcher/base_artifact_fetcher.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+"""
+:Description: Provides a base class that all Artifact Fetcher are derived from.
```
Comment (beeankha) [inline conda_recipe_manager/fetcher/base_artifact_fetcher.py:2]: Nitpicky wording change suggestion
```suggestion
:Description: Provides a base class that all ArtifactFetcher classes are derived from.
```

File: conda_recipe_manager/fetcher/base_artifact_fetcher.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+_ARTIFACT_FETCHER_FILE_ID: Final[str] = "crm_artifact_fetcher"
+
+
+class BaseArtifactFetcher(metaclass=ABCMeta):
+    """
+    Base class for all `ArtifactFetcher` classes. An `ArtifactFetcher` provides a standard set of tools to retrieve
+    bundles of source code.
+
+    Files retrieved from any artifact fetcher are stored in a secure temporary directory. That directory is deleted
+    when the Artifact Fetcher instances falls out of scope.
```
Comment (beeankha) [inline conda_recipe_manager/fetcher/base_artifact_fetcher.py:22]:
```suggestion
    when the Artifact Fetcher instance falls out of scope.
```

**Review by beeankha — APPROVED** [review]

**Pull request merged**